src/__tests__/utils/transform.tsx

```diff
@@ -1,51 +1,120 @@
+import * as Fs from 'fs';
 import * as ts from 'typescript';
-import * as fs from 'fs';
+import { Volume } from 'memfs';
 import * as path from 'path';
+import * as resolveModule from 'resolve';
+import pkg from '../../../package.json';
+import { copy, mkdirp } from './memfs';
 
 const printer = ts.createPrinter();
 
+declare module 'resolve' {
+  export interface SyncOpts extends resolveModule.Opts {
+    /** how to read files synchronously (defaults to fs.readFileSync) */
+    readFileSync?: (file: string, charset: string) => string | Buffer;
+    /** function to synchronously test whether a file exists */
+    isFile?: (file: string) => boolean;
+    /** function to synchronously test whether a directory exists */
+    isDirectory?: (directory: string) => boolean;
+  }
+}
+
 type ProgramTransformer = (program: ts.Program) => ts.TransformerFactory<ts.SourceFile>;
 
-/**
- * This creates a full project which will resolve all modules.
- * Only use this when wanting to test imports tbh. It's slow.
- */
-export const createFullTransform = (
-  programTransformer: ProgramTransformer,
-  dir: string
-) => (sources: {
+interface Sources {
   /**
    * This is the root file for the transform.
    */
   index: string;
   [key: string]: string;
-}): Promise<string> => {
+}
+
+/**
+ * This creates a full project which will resolve all modules.
+ * Only use this when wanting to test imports tbh. It's slow.
+ */
+export const createFullTransform = (programTransformer: ProgramTransformer) => (
+  sources: Sources
+): Promise<string> => {
   return new Promise((resolve, reject) => {
-    try {
-      fs.rmdirSync(`${dir}/.tmp`, { recursive: true });
-    } catch {}
-
-    fs.mkdirSync(`${dir}/.tmp`);
-    const files: string[] = [];
-    Object.keys(sources).forEach(key => {
-      const source = sources[key];
-      const filename = `${key}.tsx`;
-      const filepath = path.resolve(`${dir}/.tmp/${filename}`);
-      files.push(filepath);
-      fs.writeFileSync(filepath, source);
-    });
-
-    const [rootFile] = files;
+    const fs = (Volume.fromJSON(
+      Object.entries(sources).reduce<Record<string, string>>(
+        (acc, [name, content]) => ({
+          ...acc,
+          [`/${name}.tsx`]: content,
+        }),
+        {}
```

Comment:
if we type this can we skip having to cast it to `unknown`?

Reply:
There are some differences between `typeof Fs` and `typeof Volume`, I'll have to fiddle a bit to make good typings for the memfs utilities so this can go away. I started work on https://github.com/marionebl/ts-transformer-testing-library to that end. I'd suggest we land this and then have a look on how to use the external lib with better typings and encapsulation.

Reply:
oh awesome let's use that when you've fleshed it out more!

yeah sounds good